Monorepo integration

kitified opened the pull request:
# Overview

### Deprecates `fossa vps` commands.

The "VPS Classic" product is deprecated. Instead we're splitting into Monorepo and VSI support.

### Adds a new flag to `fossa analyze`

`experimental-enable-monorepo`: Accept a string indicating the type of monorepo scan to run, with the only currently valid value of `aosp`. If not provided, the analysis command proceeds as normal without performing a monorepo scan.
Any provided string value is passed along to the monorepo plugin for validation; spectrometer doesn't do any validation on this value.

## Acceptance criteria

* Leave `fossa analysis` invocations that are not run with `--experimental-enable-monorepo` untouched.
* When `--experimental-enable-monorepo` is provided, do not run normal analysis; instead run a scan with the monorepo plugin.

## Testing plan

* I ran `cabal run fossa -- analyze --project quick -e 'http://localhost:9578' --fossa-api-key $FOSSA_API_KEY ~/projects/quick` and validated that the command worked the same as it did before this change.
* I ran `cabal run fossa -- analyze --experimental-enable-monorepo aosp -e 'http://localhost:9578' --fossa-api-key $FOSSA_API_KEY ~/projects/aosp-mini/` and validated that the scan worked in monorepo mode.
* I tried a few different arguments in both modes and things worked as expected.

## Risks

This could potentially keep standard `fossa analyze` commands from running properly if my new [branching pattern match](https://github.com/fossas/spectrometer/compare/monorepo-integration?expand=1#diff-f2fe846f526d80e94301cb1315b5a660dea9fee7efe2fb1be098fad1112200a8R88) is not well formed. Other than that I don't think this is very risky.

## References

Pivotal tracker: https://www.pivotaltracker.com/story/show/178806134

<details><summary>Reproduced ticket text</summary>
Remove the concept of the `vps` subcommand in Spectrometer, bringing monorepo support into the standard `analyze` process.

* Add a deprecation warning to the `fossa vps` subcommand.
* Add `--experimental-enable-monorepo` to `fossa analyze`:
  * Accepts a string.
  * Forwards the string to wiggins; leave validation logic there.
* When `fossa analyze` is run with `experimental-enable-monorepo`, run `wiggins monorepo` instead of the standard analysis flow.

When running `wiggins monorepo`, pass along flags as previously implemented for `fossa vps analyze`. Additionally, pass along the value for `--experimental-enable-monorepo` to wiggins' `-type` argument.
</details>

## Checklist

_Going to work through these, but since they're doc related only the PR is ready for review now._

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] I linted and formatted (via `haskell-language-server`) any haskell files I touched in this PR.
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I linked this PR to any referenced GitHub issues, if they exist.

kitified commented on src/App/Fossa/Main.hs:
```diff
@@ -4,14 +4,15 @@ module App.Fossa.Main (
 ) where
 
-import App.Fossa.Analyze (RecordMode (..), ScanDestination (..), UnpackArchives (..), JsonOutput (..), VSIAnalysisMode (..), analyzeMain)
+import App.Fossa.Analyze (JsonOutput (..), RecordMode (..), ScanDestination (..), UnpackArchives (..), VSIAnalysisMode (..), analyzeMain)
```
Auto formatting did this.

cnr commented on src/App/Fossa/Main.hs:
```diff
@@ -84,14 +85,27 @@ appMain = do
+    AnalyzeCommand AnalyzeOptions{analyzeOutput, analyzeBranch, analyzeMetadata, monorepoAnalysisOpts = (MonorepoAnalysisOpts (Just monorepoAnalysisType)), analyzeBaseDir} -> do
+      dieOnWindows "Monorepo analysis is not supported on Windows"
+      if analyzeOutput
+        then die "Monorepo analysis does not support stdout scan destination"
```
[nit] may want to specify the `--output` flag more directly in this message

cnr commented on src/App/Fossa/Main.hs:
```diff
@@ -84,14 +85,27 @@ appMain = do
+          key <- requireKey maybeApiKey
+          let apiOpts = ApiOpts optBaseUrl key
+          let metadata = maybe analyzeMetadata (mergeFileCmdMetadata analyzeMetadata) fileConfig
+          let monorepoAnalysisOpts = MonorepoAnalysisOpts (Just monorepoAnalysisType)
```
[nit] the pattern match above should already bind `monorepoAnalysisOpts` to this value, I think(?)

kitified replied:
It doesn't seem to do so:
```
Couldn't match expected type ‘MonorepoAnalysisOpts’
              with actual type ‘AnalyzeOptions -> MonorepoAnalysisOpts’
```

And I can't access `AnalyzeOptions` because we're deconstructing it in the pattern match statement.

cnr replied:
weird. there's syntax we can use to bind the `AnalyzeOptions` as part of the pattern match:

```hs
opts@AnalyzeOptions{ etc }
```

though I'm not sure that would make this clearer. oh well

cnr commented on src/App/Fossa/Monorepo.hs:
```diff
@@ -0,0 +1,42 @@
+
+monorepoMain :: BaseDir -> MonorepoAnalysisOpts -> Severity -> ApiOpts -> ProjectMetadata -> OverrideProject -> IO ()
+monorepoMain basedir monoRepoAnalysisOpts logSeverity apiOpts projectMeta overrideProject = withDefaultLogger logSeverity $ do
+  result <- runDiagnostics $ withWigginsBinary $ monorepoScan basedir monoRepoAnalysisOpts logSeverity apiOpts projectMeta overrideProject
+  case result of
+    Left failure -> do
+      logStdout $ toText $ show $ renderFailureBundle failure
+      sendIO exitFailure
+    Right _ -> pure ()
```
[nit] use our `logWithExit_` helper here
```suggestion
  logWithExit_ $ withWigginsBinary $ monorepoScan basedir monoRepoAnalysisOpts logSeverity apiOpts projectMeta overrideProject
```

cnr reviewed (approved):
LGTM -- I don't see any mention in the code of, e.g., `experimental-monorepo-follow-symlinks` from the PR description. Is that intentional?

cnr commented:
General question: is there a suggested timeline for removing the deprecated commands?

kitified commented:
@cnr 

> I don't see any mention in the code of, e.g., experimental-monorepo-follow-symlinks from the PR description. Is that intentional?

Yes- I removed them from the code but forgot to update the PR. PR is now updated. Basically we decided to not go down that road at this time.

> General question: is there a suggested timeline for removing the deprecated commands?

There is! We currently have 3 customers using these commands: Ford, Ericsson, and Motorola. We're working on migrating them off of these commands, and once all three are migrated off we're going to remove the entire `fossa vps` subcommand.
We don't have an exact timeline but we expect we'll be able to do this by late August (the main blocker is transitioning Ericsson to our C/C++ strategy, which is pending us being ready to beta it to them).
